File: leetcode/medium/807. Max Increase to Keep City Skyline/solution.js
// https://leetcode.com/problems/max-increase-to-keep-city-skyline/
/**
 * @param {number[][]} grid
 * @return {number}
 */
function maxIncreaseKeepingSkyline(grid) {
    const len = grid.length;
    const axis = Array.from(Array(len), () => Array());
    let sum = 0;

    for (let y = 0; y < len; y += 1) {
        axis[y][0] = Math.max(...grid[y]);
        let max = 0;
        for (let x = 0; x < len; x += 1) {
            if (grid[x][y] > max) max = grid[x][y];
        }
        axis[y][1] = max;
    }

    for (let y = 0; y < len; y += 1) {
        for (let x = 0; x < len; x += 1) {
            sum += Math.min(axis[y][0], axis[x][1]) - grid[y][x];
        }
    }
    console.log(sum);
    return sum;
};

maxIncreaseKeepingSkyline([[3, 0, 8, 4], [2, 4, 5, 7], [9, 2, 6, 3], [0, 3, 1, 0]])

File: leetcode/medium/807. Max Increase to Keep City Skyline/solution.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxIncreaseKeepingSkyline(vector<vector<int>>& grid) {
        vector<int> cols;
        vector<int> rows;
        int sum = 0;

        for (auto it: grid) {
            int max = *max_element(it.begin(), it.end());
            rows.push_back(max);
        }

        int size = grid[0].size();

        for (int x = 0; x < size; x += 1) {
            int max = 0;
            for (int y = 0; y < grid.size(); y += 1) {
                if (max < grid[y][x]) max = grid[y][x];
            }
            cols.push_back(max);
        }

        for (int x = 0; x < size; x += 1) {
            for (int y = 0; y < grid.size(); y += 1) {
                sum += min(rows[x], cols[y]) - grid[x][y];
            }
        }
        cout << sum;
        return sum;
    }
};

int main() {
    Solution solution;
    vector<vector<int>> vec = {
        {3,0,8,4},
        {2,4,5,7},
        {9,2,6,3},
        {0,3,1,0}
    };

    solution.maxIncreaseKeepingSkyline(vec);
    return 0;
}